takari-builder-common/src/main/java/io/takari/builder/internal/pathmatcher/PathNormalizer.java

```diff
@@ -37,53 +49,103 @@ public String getBasedir() {
+   * under basedir.
+   */
+  boolean isBasedirOrNestedFile(String file) {
+    if (file.startsWith(basedir)) {
+      int baseLength = basedir.length();
+      if (baseLength == file.length()) {
+        return true;
+      }
+      char charL = file.charAt(baseLength);
+      return charL == SEPARATOR_CHAR || charL == File.separatorChar;
+    }
+    return false;
+  }
+
   /**
    * File path normalization implementation optimized for use with {@link PathMatcher}.
    * 
    * Normalized path may include '../' and './' special directory names, which are resolved by
    * {@link PathMatcher}.
    */
-  public String normalize(Path file) {
-    String path = file.toAbsolutePath().normalize().toString();
-    if (file.toAbsolutePath().startsWith(basedir)
-        || basepath.equals(file.toAbsolutePath().toString())) {
+  public String normalize(Path path) {
+    String file = path.toString();
+    if (isBasedirOrNestedFile(file)) {
       // avoid expensive normalize0 call if file starts with basedir prefix.
       // this is possible because basedir is normalized in the constructor and PatchMatcher
       // resolves ../ and ./ special directory names.
-      return path;
+      return FIXFS ? fixfs(file) : file;
     }
-    return normalize0(file);
+    return normalize0(path);
   }
 
   /**
-   * Returns canonical file path with '/' separator char on all platforms. Returns absolute file
-   * path if canonical path cannot be determined due to IOException.
+   * Returns canonical normalized file path, which on all platforms means:
+   * <ul>
+   * <li>the path is resolved similarly to {@link File#getCanonicalPath()}
+   * <li>the path starts with '/'
+   * <li>the path uses '/' file separator char
+   * </ul>
    */
   public static String normalize0(Path file) {
-    String normalized;
+    String canonicalPath = toCanonicalPath(file);
+    if (FIXFS) {
+      return fixfs(canonicalPath);
+    }
+    return canonicalPath;
+  }
+
+  /**
+   * Like {@link #getCanonicalPath(Path)}, but returns String.
+   */
+  private static String toCanonicalPath(Path file) {
+    try {
+      // note that Path#toRealPath() only works for existing files
+      return file.toFile().getCanonicalPath();
+    } catch (IOException e) {
+      throw new UncheckedIOException(e);
+    }
+  }
+
+  public static Path getCanonicalPath(Path file) {
     try {
-      normalized = file.toFile().getCanonicalPath();
+      // note that Path#toRealPath() only works for existing files
+      return file.toFile().getCanonicalFile().toPath();
     } catch (IOException e) {
-      normalized = file.toAbsolutePath().toString();
+      throw new UncheckedIOException(e);
     }
-    return normalized;
   }
 
   /**
    * @see #normalize0(File)
    */
   public static String normalize0(String file) {
-    return normalize0(Paths.get(file));
+    return normalize0(toPath(file));
   }
 
+  public static Path toPath(String file) {
```

Comment:
So should this method be used throughout builder framework, wherever Paths.get("...") would have been used?

Reply:
Right. We don't have to use this method when we are certain provided string can be passed to Paths.gets(...), like I believe is the case with `java.io.tmpdir` or when we use File.getCanonicalPath... but even then, it is probably better to call this always method just to be safe. I'll add this to my TODO list.